cmd/snap-confine/ns-support.c

```diff
@@ -238,8 +239,76 @@ void sc_close_ns_group(struct sc_ns_group *group)
 	free(group);
 }
 
+static dev_t find_base_snap_device(const char *base_snap_name,
+				   const char *base_snap_rev)
+{
+	// Find the backing device of the base snap.
+	// TODO: add support for "try mode" base snaps that also need
+	// consideration of the mie->root component.
+	dev_t base_snap_dev = 0;
+	char base_squashfs_path[PATH_MAX];
+	sc_must_snprintf(base_squashfs_path,
+			 sizeof base_squashfs_path, "%s/%s/%s",
+			 SNAP_MOUNT_DIR, base_snap_name, base_snap_rev);
+	struct sc_mountinfo *mi SC_CLEANUP(sc_cleanup_mountinfo) = NULL;
+	mi = sc_parse_mountinfo(NULL);
+	if (mi == NULL) {
+		die("cannot parse mountinfo of the current process");
+	}
+	bool found = false;
+	for (struct sc_mountinfo_entry * mie =
+	     sc_first_mountinfo_entry(mi); mie != NULL;
+	     mie = sc_next_mountinfo_entry(mie)) {
+		if (sc_streq(mie->mount_dir, base_squashfs_path)) {
+			base_snap_dev = MKDEV(mie->dev_major, mie->dev_minor);
+			debug("found base snap filesystem device %d:%d",
+			      mie->dev_major, mie->dev_minor);
+			// Don't break when found, we are interested in the last
+			// entry as this is the "effective" one.
+			found = true;
+		}
+	}
+	if (!found) {
+		die("cannot find device backing the base snap %s",
+		    base_snap_name);
+	}
+	return base_snap_dev;
+}
+
+static bool should_discard_current_ns(dev_t base_snap_dev)
+{
+	// Inspect the namespace and check if we should discard it.
+	//
+	// The namespace may become "stale" when the rootfs is not the same
+	// device we found above. This will happen whenever the base snap is
+	// refreshed since the namespace was first created.
+	struct sc_mountinfo_entry *mie;
+	struct sc_mountinfo *mi SC_CLEANUP(sc_cleanup_mountinfo) = NULL;
+
+	mi = sc_parse_mountinfo(NULL);
```

Comment:
If we need to reuse the info, let's hand it in from the outside instead of parsing it on every leaf function.

Reply:
We're not parsing mountinfo in any vicinity to this code but I'll do a review of when we do parse it and if we could save some calls. I'll do this in a separate PR though.